scripts/core/helpers/waitForMediaToBeReady.ts

```diff
@@ -1,37 +1,45 @@
+import {isImage, isAudio, isVideo} from './utils';
+
 /**
- * @param {Array<Node>} elements
- * @param {string} eventName
+ * @param elements An array of img, audio and video elements
  */
-const waitForEventToFireForAllElements = (elements, eventName) => new Promise((resolve) => {
-    let itemsLeftToLoad = elements.length;
+export const waitForMediaToLoad = (elements: Array<HTMLImageElement | HTMLAudioElement | HTMLVideoElement>):
+Promise<void> => new Promise((resolve) => {
+    const filteredElements = elements.filter((element) => {
+        if (isImage(element)) {
+            return element.complete === false;
+        } else if (isAudio(element) || isVideo(element)) {
+            return element.readyState < 1;
+        } else {
+            logger.error(new Error('Unexpected element type'));
+            return false;
+        }
+    });
+    let itemsLeftToLoad: number = filteredElements.length;
 
     if (itemsLeftToLoad === 0) {
         resolve();
         return;
     }
 
-    const onItemLoaded = (event) => {
+    const eventHandler = (event) => {
         itemsLeftToLoad--;
-        event.target.removeEventListener(eventName, onItemLoaded);
 
         if (itemsLeftToLoad === 0) {
             resolve();
         }
     };
 
-    elements.forEach((element) => {
-        element.addEventListener(eventName, onItemLoaded);
+    filteredElements.forEach((element) => {
+        if (isImage(element)) {
+            // check for error in image src
+            element.addEventListener('error', eventHandler, {once: true});
+        } else if (isAudio(element) || isVideo(element)) {
+            // for audio and video check for the error in source
+            const source = element.getElementsByTagName('source')[0];
+
+            source.addEventListener('error', eventHandler, {once: true});
+        }
+        element.addEventListener(isImage(element) ? 'load' : 'loadedmetadata', eventHandler, {once: true});
```

Comment:
this should be inside `else if (isAudio(element) || isVideo(element))` branch

Reply:
What's wrong with this one? :)

Reply:
Sorry, my bad. I thought the ternary was about audio/video. Since it's about image and audio/video, it makes sense to put it in respective branches.

Your approach is not wrong at the moment but you are applying similar conditions twice which is less clear. In your case, you have 4 if statements, which are almost the same, but not exactly(the ternary uses else, not else if).

It would break easily if another media type was added. It's not likely that we'll need another media type, but it's a good practice not to duplicate if statements and put everything in one place if possible.

It looks clearer if you replace ternary with `if` operator

![screenshot_7](https://user-images.githubusercontent.com/2076953/57455056-91385a00-726a-11e9-8261-e9802ba8761e.png)